### PR TITLE
Add google.auth.oauth2client - helpers for oauth2client migration

### LINKED ISSUE #38: Implement ability to convert oauth2client credentials classes
(Towards #3)

### PR DESCRIPTION
**Pull request by jonparrott**
Resolves #38 

@dhermes please let me know what you think about exposing this as `google.auth.oauth2client`, and feel free to suggest a better name (`oauth2client_compat`?)

### CODE REVIEW

File: google/auth/oauth2client.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
+# Copyright 2015 Google Inc.
```
Comment (dhermes) [inline google/auth/oauth2client.py]: 2015 on porpoise?

File: google/auth/oauth2client.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+import google.oauth2.credentials
+import google.oauth2.service_account
+
+try:
+    import oauth2client.client
+    import oauth2client.contrib.gce
+    import oauth2client.service_account
+    _HAS_OAUTH2CLIENT = True
+except ImportError:  # pragma: NO COVER
+    _HAS_OAUTH2CLIENT = False
```
Comment (dhermes) [inline google/auth/oauth2client.py]: This is an all or nothing, though isn't it conceivable that some of these imports work but the others fail?

File: google/auth/oauth2client.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+try:
+    import oauth2client.contrib.appengine
+    _HAS_APPENGINE = True
+except ImportError:
+    _HAS_APPENGINE = False
+
+
+def _convert_oauth2_credentials(credentials):
+    """Converts to :class:`google.oauth2.credentials.Credentials`."""
```
Comment (dhermes) [inline google/auth/oauth2client.py]: Input and ouput type? (The input type is more important.)

File: google/auth/oauth2client.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        client_secret=credentials.client_secret,
+        scopes=credentials.scopes)
+
+    new_credentials._expires = credentials.token_expiry
+
+    return new_credentials
+
+
+def _convert_service_account_credentials(credentials):
+    """Converts to :class:`google.oauth2.service_account.Credentials`."""
```
Comment (dhermes) [inline google/auth/oauth2client.py]: Ditto on types in docstring

File: google/auth/oauth2client.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    info = credentials.serialization_data
+    info['token_uri'] = credentials.token_uri
```
Comment (dhermes) [inline google/auth/oauth2client.py]: You are mutating an object (`credentials.serialization_data`) that you don't own

File: google/auth/oauth2client.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    return google.oauth2.service_account.Credentials.from_service_account_info(
+        info)
+
+
+def _convert_gce_app_assertion_credentials(credentials):
+    """Converts to :class:`google.auth.compute_engine.Credentials`."""
+    return google.auth.compute_engine.Credentials(
+        service_account_email=credentials.service_account_email)
```
Comment (dhermes) [inline google/auth/oauth2client.py]: Not worth copying the token or expiry?

File: google/auth/oauth2client.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+
+def _convert_appengine_app_assertion_credentials(credentials):
+    """Converts to :class:`google.auth.app_engine.Credentials`."""
+    return google.auth.app_engine.Credentials(
+        scopes=(_helpers.string_to_scopes(credentials.scope)
+                if credentials.scope is not None else None),
```
Comment (dhermes) [inline google/auth/oauth2client.py]: I'm not a big fan of these ternary expressions. This one is pretty verbose

File: google/auth/oauth2client.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    - :class:`oauth2client.contrib.gce.AppAssertionCredentials` to
+      :class:`google.auth.compute_engine.Credentials`.
+    - :class:`oauth2client.contrib.appengine.AppAssertionCredentials` to
+      :class:`google.auth.app_engine.Credentials`.
+
+    Returns:
+        google.auth.credentials.Credentials: The converted credentials.
+
+    Raises:
+        EnvironmentError: If oauth2client is not installed.
```
Comment (dhermes) [inline google/auth/oauth2client.py]: I learned something today that makes me sad.

```python
$ python -c 'print(EnvironmentError)'
<type 'exceptions.EnvironmentError'>
$ python3 -c 'print(EnvironmentError)'
<class 'OSError'>
```

File: google/auth/oauth2client.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+        ValueError: If the credentials could not be converted.
+    """
+    if not _HAS_OAUTH2CLIENT:
+        raise EnvironmentError(
+            'oauth2client is not installed and is required in order to convert'
+            'credentials.')
```
Comment (dhermes) [inline google/auth/oauth2client.py]: Why not make this an `ImportError` somewhere?

File: google/auth/oauth2client.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+    credentials_class = type(credentials)
+
+    try:
+        return _CLASS_CONVERSION_MAP[credentials_class](credentials)
+    except KeyError:
+        raise ValueError(
+            'Unable to convert {} to a google-auth credentials class.'.format(
```
Comment (dhermes) [inline google/auth/oauth2client.py]: Or:

```python
msg = _FAILURE_TEMPLATE.format(credentials_class)
raise ValueError(msg)
```

and then my eyes don't bleed from the indents

File: scripts/run_pylint.py [hunk 11]
```diff
@@ -56,7 +56,7 @@
     },
     'BASIC': {
         'method-rgx': '[a-z_][a-z0-9_]{2,40}$',
-        'function-rgx': '[a-z_][a-z0-9_]{2,40}$',
+        'function-rgx': '[a-z_][a-z0-9_]{2,45}$',
```
Comment (dhermes) [inline scripts/run_pylint.py]: Sadness. Why not disable it locally? Churn in your lint RC on a run-of-the-mill PR is a bad sign.

File: tests/test_oauth2client.py [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+    monkeypatch.setattr(_oauth2client, '_HAS_OAUTH2CLIENT', False)
+
+    with pytest.raises(EnvironmentError) as excinfo:
+        _oauth2client.convert('any value will do here')
+
+    assert excinfo.match('oauth2client is not installed')
+
+
+def test_import_has_app_engine(mock_oauth2client_gae_imports):
+    reload_module(_oauth2client)
```
Comment (dhermes) [inline tests/test_oauth2client.py]: Don't you need some kind of tear-down to put the module back the way it was before you reloaded it with `sys.modules` mocked?

**Comment by dhermes** [comment]
> please let me know what you think about exposing this as `google.auth.oauth2client`, and feel free to suggest a better name (`oauth2client_compat`?)

The very first thing I thought when I saw the first file (`docs/reference/google.auth.oauth2client.rst`) was: "aw man why is it public?"

My reflex says make it non-public.

- If you make it non-public, the name is less relevant. `_oauth2client` is fine
- If you make it public, I prefer `oauth2client_compat`

**Comment by jonparrott** [comment]
If I make it non-public, how would I expose the 'convert' function?

**Comment by dhermes** [comment]
> If I make it non-public, how would I expose the `convert` function?

Just force the non-public module to show up in the Sphinx generated docs and then put a big warning in the module docstring that the module is intentionally non-public and not intended for typical users / new users. Then let them ~~eat cake~~ import the non-public module.

**Comment by jonparrott** [comment]
@dhermes I'm rather just leave it out of the docs in that case. Cool with you?

**Comment by dhermes** [comment]
@jonparrott I suppose it's cool with me, though you should still have very good and useful docstrings (especially the module docstring).

**Comment by jonparrott** [comment]
> (especially the module docstring).

Anything you'd add to the module docstring?

**Comment by dhermes** [comment]
Essentially

> put a big warning in the module docstring that the module is intentionally non-public and not intended for typical users / new users

File: google/auth/oauth2client.py (hunk 1 above)
Comment (jonparrott) [inline google/auth/oauth2client.py]: Nope, fixed.

File: google/auth/oauth2client.py (hunk 2 above)
Comment (jonparrott) [inline google/auth/oauth2client.py]: Possibly, but only for very old versions of oauth2client. This explicitly targets the latest (and soon last) release of oauth2client.

File: google/auth/oauth2client.py (hunk 3 above)
Comment (jonparrott) [inline google/auth/oauth2client.py]: Done (for all converters)

File: google/auth/oauth2client.py (hunk 4 above)
Comment (jonparrott) [inline google/auth/oauth2client.py]: Done

File: google/auth/oauth2client.py (hunk 5 above)
Comment (jonparrott) [inline google/auth/oauth2client.py]: Yikes! Fixed.

File: google/auth/oauth2client.py (hunk 6 above)
Comment (jonparrott) [inline google/auth/oauth2client.py]: Nah, they're irrelevant.

File: google/auth/oauth2client.py (hunk 7 above)
Comment (jonparrott) [inline google/auth/oauth2client.py]: Fixed - turns out these can never be None.

File: google/auth/oauth2client.py (hunk 8 above)
Comment (jonparrott) [inline google/auth/oauth2client.py]: 😢

File: google/auth/oauth2client.py (hunk 9 above)
Comment (jonparrott) [inline google/auth/oauth2client.py]: Done.

File: google/auth/oauth2client.py (hunk 10 above)
Comment (jonparrott) [inline google/auth/oauth2client.py]: I don't like putting the error string so far away from where I throw, but fine, done.

File: scripts/run_pylint.py (hunk 11 above)
Comment (jonparrott) [inline scripts/run_pylint.py]: Disabled locally.

File: tests/test_oauth2client.py (hunk 12 above)
Comment (jonparrott) [inline tests/test_oauth2client.py]: Yeaaaah kind of. I've added a fixture that does it.

**Comment by jonparrott** [comment]
@dhermes you've approved, but I changed a lot since, care to take another look? ("No" is also acceptable as this is a completely private module)

**Pull request merged**